### DISCUSSION TITLE
How to implement custom NUnit attribute based on `TestCaseSourceAttribute`?

### DISCUSSION THREAD

**Discussion by czarls** [reqnroll/Reqnroll#246] [Q&A]
Hi all, I have been using Reqnroll for some time now along with my other C# test projects. I want to share some functionalities between them. In my other NUnit project I use custom `TestCaseParams` attribute implementing `TestCaseSourceAttribute`:
```
[AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
public class TestCaseParamsAttribute : TestCaseSourceAttribute
{
    public TestCaseParamsAttribute(int testCaseId)
        : base(typeof(DataProviders), nameof(DataProviders.GetParamObjects), new object[] { testCaseId })
    {
    }
}
```
`GetParamObjects` method, which takes argument `testCaseId` returns an array of objects which create me test parameters (in this case it retrieves Azure Test Case parameters - shared or local).

I'm wondering how to implement a new, custom plugin in Reqnroll, which creates me Scenario Outline after build, similarly to "[External Data Plugin](https://docs.reqnroll.net/latest/integrations/externaldata.html)".

Perfect solution would be a tag: @DataSourceFromTestCase:{int}

Thanks! :)

**Comment by clrudolphi** [comment]
You might try the Decorators feature in a custom Generator PlugIn: 
https://docs.reqnroll.net/latest/extend/decorators.html

Or you might try extending the External Data Plugin by creating a new IDataSourceLoader.

**Reply by czarls** [reply]
@clrudolphi thanks, Decorators with a custom Generator Plugin is what I tried but I stucked on Generator Plugin, couldn't also find any working example in Reqnroll examples, nor in Specflow (at least the one targeting .NET 6 or above).
Extending External Data Plugin is far too much now ;)
Do you know if there is any working example of custom Generator Plugin, with the feature file also to have the full flow?

**Reply by clrudolphi** [reply]
I'm not aware of any generator samples. The External Data plugin is the best available example for you to work from.

**Comment by gasparnagy** [comment]
@czarls When I cleaned up the pages https://docs.reqnroll.net/latest/extend/plugins.html#generator-plugins and https://docs.reqnroll.net/latest/extend/decorators.html, I think I have tried them out to create a generator plugin. Could you please give another try?

**Reply by czarls** [reply]
Hi @gasparnagy, yep, I went through those pages in the documentation. Decorators is clear for me however I have a feeling that Generator Plugin is missing something - that's what I understood at least from the Specflow's project code. I think I would fully understand this when having some working example - even the one from documentation.
Parts I don't get is the `*.nuspec` file and `*.csproj` configuration.
Generator Plugin and Decorator classes are clear to me.
However, it's still not working, `[NUnit.Framework.CategoryAttribute("myMethodTagDecorator")]` is being created in generated *.feature.cs file instead of ApartmentAttribute.

**Reply by gasparnagy** [reply]
I haven't really found the right place for it, but I added a simple generator plugin that uses the decorator from the docs to https://github.com/reqnroll/Reqnroll.ExploratoryTestProjects/tree/main/CustomPlugins

It also contain a test project and it works, please check the note in the [readme](https://github.com/reqnroll/Reqnroll.ExploratoryTestProjects/blob/main/CustomPlugins/README.md).
